onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/propagate_cast_ops.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+std::unordered_set<std::string> fp16_allow = {"Transpose", "Reshape", "Gather", "Split", "Relu", "Where", "Dropout"};
```

Comment:
How do we ensure we only propagate over Reshape's first input (the second input is a shape, and come as int64 type) 
similarly for Gather's index input, Where's condition, Dropout's ratio and training mode? 

I think we need to specify the input and output indices. 

e.g. 
{ Reshape, {0}, {0} }

Reply:
We only  traverse the graph up/down only on FLOAT inputs/outputs.
Without checking for any specific op type, or index of the input/output we only look at the FLOAT input/output.

Reply:
Only FLOAT/FLOAT16 casts inputs are looked at.
Cast operation on constant initializers, whether float or not are currently being skipped by PropagateCastOps.